File: Data Structures & Algorithms/generate-parentheses/submission-0.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        
        res = []

        def addParenthesis(o, c, comb):
            if c == 0 and o == 0:
                res.append(comb)
            
            if o > 0:
                addParenthesis(o - 1, c, comb + '(')
            if c > o:
                addParenthesis(o, c - 1, comb + ')')  
        addParenthesis(n,n,"")
        
        return res



            

File: Data Structures & Algorithms/insert-new-interval/submission-0.js
class Solution {
    /**
     * @param {number[][]} intervals
     * @param {number[]} newInterval
     * @return {number[][]}
     */
    insert(intervals, newInterval) {
        let res = [];

        for (let i = 0; i < intervals.length; i++) {
            // If newInterval ends before the current interval starts
            if (newInterval[1] < intervals[i][0]) {
                res.push(newInterval);
                return res.concat(intervals.slice(i));
            }
            // If newInterval overlaps with the current interval
            else if (newInterval[0] <= intervals[i][1]) {
                newInterval = [
                    Math.min(newInterval[0], intervals[i][0]),
                    Math.max(newInterval[1], intervals[i][1])
                ];
            } else {
                // If newInterval doesn't overlap, just push the current interval
                res.push(intervals[i]);
            }
        }

        // Add the final merged newInterval if it hasn't been added yet
        res.push(newInterval);
        return res;
    }
}
